### PR TITLE
Feature/json ld2

### LINKED ISSUE #17: Search arguments that span multiple indices can't use alias
Because the search will use one of the query params in the resultScan. At that point aliases no longer work.
Solution, remove aliases.

### PR DESCRIPTION
**Pull request by woutslakhorst**
- removed AliasOption fixes #17 
- added jsonLD support

### CODE REVIEW

File: collection.go [hunk 1]
```diff
@@ -36,30 +38,30 @@ var ErrNoIndex = errors.New("no index found")
...
-// documentCollection is the bucket that stores all the documents for a collection
+// documentCollection is the bucket that stores all the documents for a jsonCollection
 const documentCollection = "_documents"
 
 func documentCollectionByteRef() []byte {
 	return []byte(documentCollection)
 }
 
-// Collection defines a logical collection of documents and indices within a store.
+// Collection defines a logical jsonCollection of documents and indices within a store.
```
Comment (reinkrul) [inline collection.go:48]: errornous replacements of collection to jsonCollection?

File: examples/jsonld/main.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	fmt.Println("added docs")
+
+	query := leia.New(leia.Eq(leia.NewTermPath("http://example.com/name"), leia.MustParseScalar("name3"))).
+		And(leia.Eq(leia.NewTermPath("http://example.com/url"), leia.MustParseScalar("http://www.url3.com"))).
+		And(leia.Eq(leia.NewTermPath("http://example.com/children", "http://example.com/name"), leia.MustParseScalar("child3"))).
```
Comment (reinkrul) [inline examples/jsonld/main.go:109]: do we support binary operations on the same element of a list? Say where there's a child with name X and url Y?

File: search.go [hunk 3]
```diff
@@ -27,80 +27,142 @@ import (
...
-	Parts() []QueryPart
+// NewJSONPath creates a JSON path query: "person.path" or "person.children.#.path"
+// # is used to traverse arrays
+func NewJSONPath(path string) QueryPath {
+	return jsonPath(path)
 }
 
-type QueryPart interface {
+func (q jsonPath) Equals(other QueryPath) bool {
+	otherJSONPath, ok := other.(jsonPath)
```
Comment (reinkrul) [inline search.go:39]: is this cast needed, if you just use the equality operator below?

File: search.go [hunk 4]
```diff
@@ -27,80 +27,142 @@ import (
...
+
+// termPath represents a nested term structure (or graph path) using the fully qualified IRIs
+type termPath struct {
+	// terms represent the nested structure from highest (index 0) to lowest nesting
+	terms []string
+}
 
-	// Name returns the name that matches fieldIndexer.Name() so actually the alias or JSON path
-	Name() string
+// NewTermPath creates a QUeryPath of JSON-LD terms
```
Comment (reinkrul) [inline search.go:58]:
```suggestion
// NewTermPath creates a QUeryPath of JSON-LD terms
```
```suggestion
// NewTermPath creates a QueryPath of JSON-LD terms
```

File: search.go [hunk 5]
```diff
@@ -27,80 +27,142 @@ import (
...
+	for i, term := range tp.terms {
+		if term != otherTermPath.terms[i] {
+			return false
+		}
+	}
+	return true
+}
+
+type QueryPart interface {
+	IRIComparable
```
Comment (reinkrul) [inline search.go:104]: do we use IRIs for JSON paths as well?

File: store.go [hunk 6]
```diff
@@ -23,55 +23,101 @@ import (
...
+		s.collections[name] = c
 	}
 
-	return st, nil
+	return c
 }
 
-func (s *store) Collection(name string) Collection {
+func (s *store) JSONLDCollection(name string) Collection {
 	c, ok := s.collections[name]
```
Comment (reinkrul) [inline store.go:113]: what if the collection was created earlier as JSON collection, or vice versa?

File: collection.go [hunk 7]
```diff
@@ -366,23 +362,105 @@ func (c *collection) Get(key Reference) (Document, error) {
...
+		if value, ok := expanded["@value"]; ok {
+			return []Scalar{MustParseScalar(value)}
+		}
+		if id, ok := expanded["@id"]; ok {
+			return []Scalar{MustParseScalar(id)}
+		}
+		if list, ok := expanded["@list"]; ok {
+			castList := list.([]interface{})
+			return valuesFromSliceAtPath(castList, termPath)
+		}
```
Comment (reinkrul) [inline collection.go:444]: what about `@set` then?

File: collection.go [hunk 8]
```diff
@@ -366,23 +362,105 @@ func (c *collection) Get(key Reference) (Document, error) {
...
+	}
+
+	if value, ok := expanded[termPath.Head()]; ok {
+		// the value should now be a slice
+		next, ok := value.([]interface{})
+		if !ok {
+			return nil
```
Comment (reinkrul) [inline collection.go:451]: can this actually happen? Wouldn't it be a bug?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: collection.go (hunk 7 above)
Comment (woutslakhorst) [reply collection.go:444]: `@set` translates into a list of `@value`s

File: search.go (hunk 3 above)
Comment (woutslakhorst) [reply search.go:39]: it's not

File: examples/jsonld/main.go (hunk 2 above)
Comment (woutslakhorst) [reply examples/jsonld/main.go:109]: no, lower level boolean logic is not supported

File: store.go (hunk 6 above)
Comment (woutslakhorst) [reply store.go:113]: Then you're screwed.

File: collection.go (hunk 8 above)
Comment (woutslakhorst) [reply collection.go:451]: It shouldn't happen. Better to return nothing then to crash the entire program

File: search.go [hunk 9]
```diff
@@ -27,80 +27,142 @@ import (
...
+
+	return q == otherJSONPath
+}
+
+// QueryPath is the interface for the query path given in queries
+type QueryPath interface {
+	Equals(other QueryPath) bool
+}
+
+// termPath represents a nested term structure (or graph path) using the fully qualified IRIs
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply search.go]: renamed to `iriPath`. `term` is actually used for the shorthand notation in JSON-LD (compact form)

**Review by reinkrul — APPROVED** [review]

**Pull request merged**